ssl/ssl_key_share.cc

```diff
@@ -294,6 +295,233 @@ class CECPQ2KeyShare : public SSLKeyShare {
+    size_t secret_bytes_written = secret_len_;
+    if (!EVP_PKEY_encapsulate(ctx_.get(), ciphertext.data(),
+                              &ciphertext_bytes_written, shared_secret.data(),
+                              &secret_bytes_written) ||
+                              ciphertext_bytes_written != ciphertext_len_ ||
+                              secret_bytes_written != secret_len_ ||
+                              !CBB_add_bytes(out_public_key, ciphertext.data(), ciphertext_len_)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // ...we retain the shared secret for our own use.
+    *out_secret = std::move(shared_secret);
+
+    // Success; clear the alert
+    *out_alert = 0;
+    return true;
+  }
+
+  bool Finish(Array<uint8_t> *out_secret, uint8_t *out_alert,
+              Span<const uint8_t> peer_key) override {
+    // Basic nullptr checks
+    if (!out_alert) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+      return false;
+    }
+
+    // Set alert to internal error by default
+    *out_alert = SSL_AD_INTERNAL_ERROR;
+
+    if (!out_secret) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+      return false;
+    }
+
+    // ctx_ should have been initialized by a previous call to Offer()
+    if (!ctx_) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+      return false;
+    }
+
+    // Retrieve the lengths of the ciphertext and shared secret
+    if (!EVP_PKEY_encapsulate(ctx_.get(), nullptr, &ciphertext_len_, nullptr,
+                              &secret_len_)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // Ensure that peer_key is valid
+    if (!peer_key.data() || peer_key.size() != ciphertext_len_) {
+      *out_alert = SSL_AD_DECODE_ERROR;
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // |peer_key| is the ciphertext, encapsulating the shared secret,
+    // that the peer generated using our public key. We use our private
+    // key to decapsulate it, and retain the shared secret by writing
+    // it to |out_secret|.
+    uint8_t *ciphertext = (uint8_t *)peer_key.begin();
+    if (!out_secret->Init(secret_len_)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_MALLOC_FAILURE);
+      return false;
+    }
+
+    size_t secret_bytes_written = secret_len_;
+    if (!EVP_PKEY_decapsulate(ctx_.get(), out_secret->data(),
+                              &secret_bytes_written, ciphertext,
+                              ciphertext_len_) ||
+                              secret_bytes_written != secret_len_) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // Success; clear the alert
+    *out_alert = 0;
+    return true;
+  }
+
+ private:
+  int nid_;
+  uint16_t group_id_;
+  UniquePtr<EVP_PKEY_CTX> ctx_;
+  size_t public_key_len_;
+  size_t ciphertext_len_;
+  size_t secret_len_;
```

Comment:
do we need these three?

Reply:
Probably not. I don't actually see any place where I'm relying on their state between calls.

Reply:
Removed the member variables from the class definition